### PR TITLE
[CI] Start testing examples

### PR DESCRIPTION
**Pull request by czechboy0**
### Motivation

We're expanding the Examples section and want to verify that they continue to work as the project evolves.

That means that they successfully build, and for those with tests, that the tests are passing.

### Modifications

Added a script that is now triggered by the new `examples` pipeline.

It goes through all the packages in Examples and builds them, and for those with a Tests directory, also tests them.

### Result

More confidence that our Examples are working.

### Test Plan

Ran the test on macOS, ran it in docker-compose, all passed, so did the new CI. Verified in CI logs that the script is actually running and the two existing examples were exercised.

### CODE REVIEW

File: scripts/test-examples.sh [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+##===----------------------------------------------------------------------===##
+set -euo pipefail
+
+log() { printf -- "** %s\n" "$*" >&2; }
+error() { printf -- "** ERROR: %s\n" "$*" >&2; }
+fatal() { error "$@"; exit 1; }
+
+log "Checking required executables..."
+SWIFT_BIN=${SWIFT_BIN:-$(command -v swift || xcrun -f swift)} || fatal "SWIFT_BIN unset and no swift on PATH"
+JQ_BIN=${JQ_BIN:-$(command -v jq)} || fatal "JQ_BIN unset and no jq on PATH"
```
Comment (simonjbeaumont) [inline scripts/test-examples.sh:23]:
```suggestion
```

File: scripts/test-examples.sh [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+for EXAMPLE_PACKAGE_PATH in $(find "${EXAMPLES_PACKAGE_PATH}" -maxdepth 2 -name Package.swift -type f | xargs dirname); do
+
+    EXAMPLE_PACKAGE_NAME=$(basename "${EXAMPLE_PACKAGE_PATH}")
+    EXAMPLE_COPY_DIR="${TMP_DIR}/${EXAMPLE_PACKAGE_NAME}"
+    log "Copying example ${EXAMPLE_PACKAGE_NAME} to ${EXAMPLE_COPY_DIR}"
+    cp -R "${EXAMPLE_PACKAGE_PATH}" "${EXAMPLE_COPY_DIR}"
+
+    log "Overriding dependency in ${EXAMPLE_PACKAGE_NAME} to use ${PACKAGE_PATH}"
+    swift package --package-path "${EXAMPLE_COPY_DIR}" \
```
Comment (simonjbeaumont) [inline scripts/test-examples.sh:40]:
```suggestion
    "${SWIFT_BIN}" package --package-path "${EXAMPLE_COPY_DIR}" \
```

File: scripts/test-examples.sh [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        edit swift-openapi-generator --path "${PACKAGE_PATH}"
+
+    log "Building example package: ${EXAMPLE_PACKAGE_NAME}"
+    swift build --package-path "${EXAMPLE_COPY_DIR}"
```
Comment (simonjbeaumont) [inline scripts/test-examples.sh:44]:
```suggestion
    "${SWIFT_BIN}" build --package-path "${EXAMPLE_COPY_DIR}"
```

File: scripts/test-examples.sh [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+JQ_BIN=${JQ_BIN:-$(command -v jq)} || fatal "JQ_BIN unset and no jq on PATH"
+
+CURRENT_SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+REPO_ROOT="$(git -C "${CURRENT_SCRIPT_DIR}" rev-parse --show-toplevel)"
+TMP_DIR=$(/usr/bin/mktemp -d -p "${TMPDIR-/tmp}" "$(basename "$0").XXXXXXXXXX")
+
+PACKAGE_PATH=${PACKAGE_PATH:-${REPO_ROOT}}
+EXAMPLES_PACKAGE_PATH="${PACKAGE_PATH}/Examples"
+
+for EXAMPLE_PACKAGE_PATH in $(find "${EXAMPLES_PACKAGE_PATH}" -maxdepth 2 -name Package.swift -type f | xargs dirname); do
```
Comment (simonjbeaumont) [inline scripts/test-examples.sh:32]:
```suggestion
for EXAMPLE_PACKAGE_PATH in $(find "${EXAMPLES_PACKAGE_PATH}" -maxdepth 2 -name Package.swift -type f -print0 | xargs -0 dirname); do
```

File: scripts/test-examples.sh [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+    EXAMPLE_PACKAGE_NAME=$(basename "${EXAMPLE_PACKAGE_PATH}")
```
Comment (simonjbeaumont) [inline scripts/test-examples.sh:34]:
```suggestion
    EXAMPLE_PACKAGE_NAME="$(basename "${EXAMPLE_PACKAGE_PATH}")"
```

File: scripts/test-examples.sh [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    EXAMPLE_COPY_DIR="${TMP_DIR}/${EXAMPLE_PACKAGE_NAME}"
+    log "Copying example ${EXAMPLE_PACKAGE_NAME} to ${EXAMPLE_COPY_DIR}"
+    cp -R "${EXAMPLE_PACKAGE_PATH}" "${EXAMPLE_COPY_DIR}"
```
Comment (simonjbeaumont) [inline scripts/test-examples.sh:37]: Fun. In all these years I never knew `-R` is the same as `-r`.

**Comment by czechboy0** [comment]
Thanks @simonjbeaumont, applied all your suggestions 🙏

**Review by simonjbeaumont — APPROVED** [review]
LGTM. If the pipeline passes, :shipit:

**Comment by czechboy0** [comment]
> LGTM. If the pipeline passes, :shipit:

It did 😎

**Pull request merged**